### PR TITLE
Beetle / Bug Enemy improved & invencibility time to player

### PR DESCRIPTION
**Pull request by VBoschMatas**
Beetle overall improvements.

- Idle has  a Patrol movement
- They now Spawn, being inmune to all damage while doing so
- They do "Player like" attacks, their attacks can be avoided by dashing or moving out of the area.

 The player has a short time of inmunity after being damaged
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

Some extra changes regarding GameObject activation order. Children first, then parent.

### CODE REVIEW

**Comment by monikamnc** [comment]
Credits to @VBoschMatas and @miquelmiro3 that helped my in my tiny commit :_) <3

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 1]
```diff
@@ -352,7 +481,12 @@ void Hachiko::Scripting::EnemyController::CheckState()
 	case BugState::MOVING_BACK:
 		animation->SendTrigger("isMoving");
 		break;
-
+	case BugState::PATROL:
+		animation->SendTrigger("isMoving");
+		break;
+	//case BugState::SPAWNING:
+	//	  animation->SendTrigger("isSpawning");
+	//	  break;
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:489]: Let's delete this commented lines.

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 2]
```diff
@@ -183,40 +264,58 @@ void Hachiko::Scripting::EnemyController::RegisterHit(int damage, float3 directi
...
 	}
-
-	_state = BugState::ATTACKING;
 	
 	if (_is_ranged_attack) 
 	{
-		math::float3 forward = _player->GetTransform()->GetGlobalPosition() - game_object->GetTransform()->GetGlobalPosition();
-		forward = forward.Normalized();
+		//math::float3 forward = _player->GetTransform()->GetGlobalPosition() - game_object->GetTransform()->GetGlobalPosition();
+		//forward = forward.Normalized();
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:297]: We can delete these if they won't be used in the follow-up pr for worm.

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 3]
```diff
@@ -183,40 +264,58 @@ void Hachiko::Scripting::EnemyController::RegisterHit(int damage, float3 directi
...
 	}
-	_attack_cooldown -= Time::DeltaTime();
-	_attack_cooldown = _attack_cooldown < 0.0f ? 0.0f : _attack_cooldown;
 
-	if (_is_ranged_attack)
-	{
-		transform->LookAtTarget(_player->GetTransform()->GetGlobalPosition());
+	//if (_is_ranged_attack)
+	//{
+	//	transform->LookAtTarget(_player->GetTransform()->GetGlobalPosition());
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:274]: Let's delete these.

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 4]
```diff
@@ -183,40 +264,58 @@ void Hachiko::Scripting::EnemyController::RegisterHit(int damage, float3 directi
...
 
-		// Make the enemy stop (quick fix)
-		ComponentAgent* agc = game_object->GetComponent<ComponentAgent>();
-		agc->SetTargetPosition(game_object->GetTransform()->GetGlobalPosition());
+	//	// Make the enemy stop (quick fix)
+	//	ComponentAgent* agc = game_object->GetComponent<ComponentAgent>();
+	//	agc->SetTargetPosition(game_object->GetTransform()->GetGlobalPosition());
+	//}
+	//
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:280]: We can delete these if they are not needed.

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 5]
```diff
@@ -154,18 +219,34 @@ void Hachiko::Scripting::EnemyController::OnUpdate()
...
 }
 
-Hachiko::Scripting::BugState Hachiko::Scripting::EnemyController::GetState() const
+void Hachiko::Scripting::EnemyController::IdleController()
 {
-	return _state;
-}
+	if (_state == BugState::IDLE)
+	{
+		_idle_cooldown -= Time::DeltaTime();
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:226]:
```suggestion
		_idle_cooldown -= Time::DeltaTime();
		_idle_cooldown = _idle_cooldown < 0.0f ? 0.0f : _idle_cooldown;
```

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 6]
```diff
@@ -31,28 +32,26 @@ Hachiko::Scripting::EnemyController::EnemyController(GameObject* game_object)
...
-void Hachiko::Scripting::EnemyController::OnStart()
-{
+	//TODO: Get Player and CombatManager from Scene script once its merged
+	_combat_manager = game_object->scene_owner->GetRoot()->GetFirstChildWithName("CombatManager")->GetComponent<CombatManager>();
+	//_attack_range = 1.5f;
+	_combat_stats = game_object->GetComponent<Stats>();
+	_combat_stats->_attack_power = 1;
+	//_combat_stats->_attack_cd = _is_ranged_attack ? 2.0f : 1.0f;
+	//_combat_stats->_move_speed = 4;
+	//_combat_stats->_max_hp = 2;
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:49]:
```suggestion
	_combat_stats = game_object->GetComponent<Stats>();
	_combat_stats->_attack_power = 1;
```

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 7]
```diff
@@ -71,42 +70,103 @@ void Hachiko::Scripting::EnemyController::OnStart()
 		_spawn_rot = transform->GetGlobalRotation();
 	}
 	animation->StartAnimating();
+	srand((unsigned)time(NULL));
+}
+
+void Hachiko::Scripting::EnemyController::OnStart()
+{
+
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:78]:
```suggestion
```

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 8]
```diff
@@ -121,27 +181,32 @@ void Hachiko::Scripting::EnemyController::OnUpdate()
...
+	if (IsAttacking())
 	{
 		_attack_animation_timer += Time::DeltaTime();
-		
+
 		if (_attack_animation_timer >= _attack_animation_duration)
 		{
 			_attack_animation_timer = 0.0f;
 			_state = BugState::IDLE;
+			_idle_cooldown = 2.0f;
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:201]: Maybe we can make this cooldown value settable through the editor?

File: Gameplay/entities/enemies/EnemyController.cpp [hunk 9]
```diff
@@ -361,14 +495,16 @@ void Hachiko::Scripting::EnemyController::CheckState()
...
-	_combat_stats = game_object->GetComponent<Stats>();
-	_combat_stats->_attack_power = 1;
-	_combat_stats->_attack_cd = _is_ranged_attack ? 2.0f : 1.0f;
-	_combat_stats->_move_speed = 4;
-	_combat_stats->_max_hp = 2;
+	//_combat_stats = game_object->GetComponent<Stats>();
+	//_combat_stats->_attack_power = 1;
+	//_combat_stats->_attack_cd = _is_ranged_attack ? 2.0f : 1.0f;
+	//_combat_stats->_move_speed = 4;
+	//_combat_stats->_max_hp = 2;
```
Comment (baransrc) [inline Gameplay/entities/enemies/EnemyController.cpp:502]:
```suggestion
```

File: Gameplay/entities/player/CombatManager.cpp [hunk 10]
```diff
@@ -563,6 +585,29 @@ bool Hachiko::Scripting::CombatManager::ProcessObstaclesOBB(const OBB& attack_bo
 	return hit;
 }
 
+bool Hachiko::Scripting::CombatManager::ProcessPlayerOBB(const OBB& attack_box, const AttackStats& attack_stats)
+{
+
```
Comment (baransrc) [inline Gameplay/entities/player/CombatManager.cpp:590]:
```suggestion
```

File: Gameplay/entities/player/PlayerController.cpp [hunk 11]
```diff
@@ -870,8 +876,9 @@ void Hachiko::Scripting::PlayerController::PickupParasite(const float3& current_
 
 void Hachiko::Scripting::PlayerController::RegisterHit(float damage_received, bool is_heavy, float3 direction)
 {
-	if (_god_mode)	return;
+	if (_god_mode || _invulnerability_time_remaining > 0.0f)	return;
```
Comment (baransrc) [inline Gameplay/entities/player/PlayerController.cpp:879]:
```suggestion
	if (_god_mode || _invulnerability_time_remaining > 0.0f)
	{
	    return;
	}
```

**Review by baransrc — CHANGES_REQUESTED** [review]
Good f**👑 job guys!! After bunch of fixes we are good to merge 💯

**Review by alvarosoppelsa — APPROVED** [review]

**Review by miquelmiro3 — APPROVED** [review]
All good! Just a couple of things, i think that the explosive crystal is not effecting enemies, the attack areas look good but a little weird maybe they are too fast, the idle patrol -> idle -> patrol behaviour is too fast (or maybe the end of the patrol is not detected properly).
Also are you going to make the player getting agro when shooting an enemy?

**Comment by VBoschMatas** [comment]
> All good! Just a couple of things, i think that the explosive crystal is not effecting enemies, the attack areas look good but a little weird maybe they are too fast, the idle patrol -> idle -> patrol behaviour is too fast (or maybe the end of the patrol is not detected properly). Also are you going to make the player getting agro when shooting an enemy?

The explosive crystal thing I guess its something that's already in development because I haven't changed anything, but I was thinking about changing a bit the crystal behaviour to break other crystals (its a super small change, but to keep it separated), so I'll fix that there.
For the other things, sure gonan fix 'em asap

**Review by baransrc — APPROVED** [review]

**Comment by Bernatmago** [comment]
image
I didnt check the pr yet but have a look at this just in case

**Pull request merged**